crates/ty_server/tests/e2e/pull_diagnostics.rs

```diff
@@ -27,3 +30,150 @@ def foo() -> str:
+def foo() -> str:
+    return 42  # Error appears
+";
+
+    // File C: Initially has error, will be fixed (removed diagnostic)
+    let file_c = SystemPath::new("src/fixed_error.py");
+    let file_c_content_v1 = "\
+def foo() -> str:
+    return 42  # Error initially
+";
+    let file_c_content_v2 = "\
+def foo() -> str:
+    return \"fixed\"  # Error removed
+";
+
+    // File D: Has error that changes content (changed diagnostic)
+    let file_d = SystemPath::new("src/changed_error.py");
+    let file_d_content_v1 = "\
+def foo() -> str:
+    return 42  # First error: expected str, got int
+";
+    let file_d_content_v2 = "\
+def foo() -> int:
+    return \"hello\"  # Different error: expected int, got str
+";
+
+    let global_options = ClientOptions::default().with_diagnostic_mode(DiagnosticMode::Workspace);
+
+    let mut server = TestServerBuilder::new()?
+        .with_workspace(
+            workspace_root,
+            ClientOptions::default().with_diagnostic_mode(DiagnosticMode::Workspace),
+        )?
+        .with_global_options(global_options)
+        .with_file(file_a, file_a_content)?
+        .with_file(file_b, file_b_content_v1)?
+        .with_file(file_c, file_c_content_v1)?
+        .with_file(file_d, file_d_content_v1)?
+        .enable_pull_diagnostics(true)
+        .build()?
+        .wait_until_workspaces_are_initialized()?;
+
+    server.open_text_document(file_a, &file_a_content, 1);
+
+    // First request with no previous result IDs
+    let first_response = server.workspace_diagnostic_request(None)?;
+    insta::assert_debug_snapshot!("workspace_diagnostic_initial_state", first_response);
+
+    // Extract result IDs from the first response
+    let previous_result_ids = match first_response {
+        WorkspaceDiagnosticReportResult::Report(report) => {
+            report.items.into_iter().filter_map(|item| {
+                match item {
+                    WorkspaceDocumentDiagnosticReport::Full(full_report) => {
+                        let result_id = full_report.full_document_diagnostic_report.result_id?;
+                        Some(PreviousResultId {
+                            uri: full_report.uri,
+                            value: result_id,
+                        })
+                    }
+                    WorkspaceDocumentDiagnosticReport::Unchanged(_) => {
+                        // Shouldn't happen in the first request
+                        None
+                    }
+                }
+            })
+        }
+        WorkspaceDiagnosticReportResult::Partial(_) => {
+            panic!("The first response must be a full report");
+        }
+    }
+    .collect();
+
+    // Make changes to files B, C, and D (leave A unchanged)
+    // Need to open files before changing them
+    server.open_text_document(file_b, &file_b_content_v1, 1);
+    server.open_text_document(file_c, &file_c_content_v1, 1);
+    server.open_text_document(file_d, &file_d_content_v1, 1);
```

Comment:
Should we close these files after making these changes? We could have a helper method on `TestServer` that does this for us but it also isn't too verbose here either.

Reply:
I think that makes sense for a test, testing the workspace diagnostic mode. I don't think it's necessary here and it only complicates the test (it's also sort of tested already by the fact that the initial run only opens 1 file)